### PR TITLE
OCP upgrade with force or notRecommended options

### PR DESCRIPTION
**Pull request by Santhoshairody**
Performs OCP upgrade to a new version with force option when the upgrade path is not  available or, to "not recommended" version when path is available but upgrading is a risk 
**What this PR does / why we need it**:
Currently we need to do upgrade builds manually when no specific path available or only not recommended path is available.
This helps to do end-to-end upgrade with multiple hops
**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

https://jenkins.pwx.dev.purestorage.com/job/Torpedo/view/Torpedo%20NextPX/job/tp-ocp-upgrade/226/console
Above test logs where I did upgrade from 4.10.50 and 4.11.25. (Upgrade with force)
I will work on upgrading to  not recommended path and post you the result here

### CODE REVIEW

File: drivers/scheduler/openshift/openshift.go [hunk 1]
```diff
@@ -330,13 +332,66 @@ spec:
...
+		Url:      downloadURL,
+		Content:  "application/json",
+		Body:     nil,
+		Insecure: true,
+	}
+	log.Infof("URL %s", downloadURL)
+	content, err := netutil.DoRequest(request)
+	if err != nil {
+		log.Errorf("Failed to get Get content from  %s, error %v", downloadURL, err)
+		return "", err
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:350]: Don't log error here, just return the whole verbose error and log it where you call this func if needed
```
return "", fmt.Errorf("Failed to get Get content from %s, error %v", downloadURL, err)
```

File: drivers/scheduler/openshift/openshift.go [hunk 2]
```diff
@@ -330,13 +332,66 @@ spec:
 	return err
 }
 
+// getImageSha get Imgae sha
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:335]: nit: get Image sha

File: pkg/netutil/netutil.go [hunk 3]
```diff
@@ -1,11 +1,60 @@
...
+		return nil, fmt.Errorf("Failed to send request, Err: %v %v", resp, err)
+	}
+	defer resp.Body.Close()
+
+	htmlData, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read response: %+v", resp.Body)
+	}
+
+	//log.Debug(string(htmlData))
```
Comment (nikolaypopov) [inline pkg/netutil/netutil.go:58]: Remove this comment

File: drivers/scheduler/openshift/openshift.go [hunk 4]
```diff
@@ -330,13 +332,66 @@ spec:
...
 	var output []byte
 	var err error
-
+	var shaName string
 	args := []string{"adm", "upgrade", fmt.Sprintf("--to=%s", upgradeVersion)}
 	t := func() (interface{}, bool, error) {
 		if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+			// If OC upgrade failed for reasons like not recommended upgrade or path does not exist.
+			forceUpgrade := "specify --to-image"
+			notRecommended := "is not one of the recommended updates, but is available"
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:372]: nit: Might want to grep for specific words that will always come up in not recommended upgrades, looking for a long sentence might not be a case for all Openshift versions, but I don't know for sure, in my opinion the smaller the string you are looking for the better

File: drivers/scheduler/openshift/openshift.go [hunk 5]
```diff
@@ -330,13 +332,66 @@ spec:
...
+}
+
 func startUpgrade(upgradeVersion string) error {
 	var output []byte
 	var err error
-
+	var shaName string
 	args := []string{"adm", "upgrade", fmt.Sprintf("--to=%s", upgradeVersion)}
 	t := func() (interface{}, bool, error) {
 		if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:369]: Add a debug log for what exact command you are running, will be easier to debug later when it fails. We might want to create our own `execCommand()` func that will use `exec.Command()`, but we can add all the verbosity and levels to it

File: drivers/scheduler/openshift/openshift.go [hunk 6]
```diff
@@ -330,13 +332,66 @@ spec:
...
+					return output, true, fmt.Errorf("failed to start upgrade due to %s. cause: %v", string(output), err)
+				}
+				log.Infof("Upgrded successful with notRecommended")
+			} else if strings.Contains(string(output), forceUpgrade) {
+				log.Infof("Upgrade with force call ")
+				if shaName, err = getImageSha(upgradeVersion); err != nil {
+					return "", false, err
+				}
+				imagePath := fmt.Sprintf("--to-image=quay.io/openshift-release-dev/ocp-release@%s", shaName)
+				log.Infof("Full command: %s", imagePath)
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:386]: nit: Remove this line once you implement our own exec command func

File: drivers/scheduler/openshift/openshift.go (hunk 4 above)
Comment (Santhoshairody) [reply drivers/scheduler/openshift/openshift.go:374]: > nit: Might want to grep for specific words that will always come up in not recommended upgrades, looking for a long sentence might not be a case for all Openshift versions, but I don't know for sure, in my opinion the smaller the string you are looking for the better
Tried to check specific words , but I feel keeping this  long sentence will help to improve readability of the code.

**Comment by Santhoshairody** [comment]
For NotRecommended upgrade test logs: https://jenkins.pwx.dev.purestorage.com/job/Torpedo/view/Torpedo%20NextPX/job/tp-ocp-upgrade/250/console
For Force upgrade logs  : https://jenkins.pwx.dev.purestorage.com/job/Torpedo/view/Torpedo%20NextPX/job/tp-ocp-upgrade/252/console

**Review by nikolaypopov — COMMENTED** [review]
Hey overall logic looks good to me, I saw both jobs, let's fix nits like extra spaces and extra not needed logs, I saw the output and its a little too much of repetitive stuff that is just confusing rather than telling us more info. If we already printing all the output in the `ExecTorpedo`, there is no need to print it again, if we not printing inside the `ExecTorpedo` by default, which is I think is the correct way, we should be printing where we call the function, but lets not print too much of everything, if we do not expect error, lets not print err or stdErr, just print out and so on, we can quickly chat tomorrow morning about the output and please make clusterVersion check wrap that we talked about earlier today

**Comment by Santhoshairody** [comment]
https://jenkins.pwx.dev.purestorage.com/job/Torpedo/view/Torpedo%20NextPX/job/tp-ocp-upgrade/260/consoleFull

File: drivers/scheduler/openshift/openshift.go [hunk 7]
```diff
@@ -323,26 +326,88 @@ spec:
 		if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
 			return nil, true, fmt.Errorf("failed to select channel due to %s. cause: %v", string(output), err)
 		}
-		log.Info(string(output))
+		log.Info(output)
+		log.Debugf("Wait 5 minutes to set to new channel")
+		time.Sleep(5 * time.Minute)
```
[the remark this answers is not in the document]
Comment (Santhoshairody) [reply drivers/scheduler/openshift/openshift.go]: Here are the improvements I did:
1. Removed print log on called methods , now you should be having less logs printed.
2. I want to print stdErr from method  where we call method as this provides some additional warning info which is necessary to log.
     "02:53:46.121 2023-02-15 08:32:04 +0000:[DEBUG log.go:#246]  StdError warning: The requested upgrade image is not one of the available updates.You have used --allow-explicit-upgrade for the update to proceed anyway
02:53:46.121 warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures."
3. Wait time has been replaced by "waitForNewGenertionNumber".

File: drivers/scheduler/openshift/openshift.go [hunk 8]
```diff
@@ -323,26 +326,88 @@ spec:
...
+	}
+	//Convert the body to type string
+	contentInString := string(content)
+	parts := strings.Split(contentInString, "\n")
+	for _, a := range parts {
+		if strings.Contains(a, "Image Digest:") {
+			return strings.Split(a, "`")[1], nil
+		}
+	}
+	return "", fmt.Errorf("Failed to find Image sha: in  %s", downloadURL)
```
[the remark this answers is not in the document]
Comment (Santhoshairody) [reply drivers/scheduler/openshift/openshift.go:400]: Removed all extra space

File: pkg/osutils/osutils.go [hunk 9]
```diff
@@ -64,13 +64,46 @@ func FileEmpty(filename os.FileInfo) bool {
 	return false
 }
 
+func ExecTorpedo(command string, arguments ...string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (Santhoshairody) [reply pkg/osutils/osutils.go]: Modified name accordingly.

File: drivers/scheduler/openshift/openshift.go [hunk 10]
```diff
@@ -323,26 +326,88 @@ spec:
...
+			notRecommended := "is not one of the recommended updates, but is available"
+			if strings.Contains(string(stdErr), notRecommended) {
+				args = []string{"adm", "upgrade", fmt.Sprintf("--to=%s", upgradeVersion), "--allow-not-recommended"}
+				log.Infof("Retrying with not recommended options")
+				output, stdErr, err = osutils.ExecTorpedo("oc", args...)
+				if err != nil {
+					return output, true, fmt.Errorf("failed to start upgrade due to %s, cause: %v ", stdErr, err)
+				}
+				log.Debugf("Upgrded called  with notRecommended %s", output)
+				log.Debugf("StdError %s", stdErr)
```
[the remark this answers is not in the document]
Comment (Santhoshairody) [reply drivers/scheduler/openshift/openshift.go:421]: 1. Removed print log on called methods , now you should be having less logs printed.
2. I want to print stdErr from method where we call method as this provides some additional warning info which is necessary to log.
"02:53:46.121 2023-02-15 08:32:04 +0000:[DEBUG log.go:https://github.com/portworx/torpedo/pull/246] StdError warning: The requested upgrade image is not one of the available updates.You have used --allow-explicit-upgrade for the update to proceed anyway
02:53:46.121 warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures."
  Hence log.Debugf("StdError %s", stdErr) is required after exec command

File: drivers/scheduler/openshift/openshift.go [hunk 11]
```diff
@@ -323,29 +363,86 @@ spec:
...
+		output, stdErr, err := osutils.ExecTorpedoShell("oc", args...)
+		if err != nil {
+			forceUpgrade := "specify --to-image"
+			notRecommended := "is not one of the recommended updates, but is available"
+			if strings.Contains(string(stdErr), notRecommended) {
+				args = []string{"adm", "upgrade", fmt.Sprintf("--to=%s", upgradeVersion), "--allow-not-recommended"}
+				log.Infof("Retrying upgrade with --allow-not-recommended option")
+				output, stdErr, err = osutils.ExecTorpedoShell("oc", args...)
+				if err != nil {
+					return output, true, fmt.Errorf("failed to start upgrade due to %s, cause: %v ", stdErr, err)
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:418]: nit: get rid of extra space

File: drivers/scheduler/openshift/openshift.go [hunk 12]
```diff
@@ -304,16 +309,51 @@ func getClientVersion() (string, error) {
...
-
 	channel := ""
 	if channel, err = getChannel(version); err != nil {
 		return err
 	}
+	beforeGenNumInt, err := getGenerationNumber()
+	if err != nil {
+		return fmt.Errorf("Failed to convert generator number from string to int : cause: %v", err)
+	}
+	log.Infof("Generation number before select channel: %d ", beforeGenNumInt)
```
Comment (nikolaypopov) [inline drivers/scheduler/openshift/openshift.go:355]: nit: same here get rid of extra space

**Review by nikolaypopov — APPROVED** [review]
Fix nits and then it lgtm

File: drivers/scheduler/openshift/openshift.go [hunk 13]
```diff
@@ -304,16 +309,51 @@ func getClientVersion() (string, error) {
...
+		if err != nil {
+			return nil, true, fmt.Errorf("Failed to convert generator number from string to int : cause: %v", err)
+		}
+		if newGenNumInt == currentGenNumber {
+			return nil, false, fmt.Errorf("Generation number has not changed yet: %d", currentGenNumber)
+		}
+		log.Debugf("Set channel spec has been updated: Generation number %d", newGenNumInt)
+		return nil, true, nil
+	}
+	_, err = task.DoRetryWithTimeout(t, 10*time.Minute, 5*time.Second)
```
Comment (jainsmit) [inline drivers/scheduler/openshift/openshift.go:340]: Define constant for 10 Minutes timeout ..

**Review by jainsmit — APPROVED** [review]
One minor comment. LGTM !!

**Pull request merged**